libcloudforensics/providers/aws/internal/iam.py

```diff
@@ -0,0 +1,212 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""AWS IAM Functionality"""
+
+import os
+
+from typing import TYPE_CHECKING, Optional
+from libcloudforensics import errors, logging_utils
+from libcloudforensics.providers.aws.internal import common
+
+if TYPE_CHECKING:
+  # TYPE_CHECKING is always False at runtime, therefore it is safe to ignore
+  # the following cyclic import, as it it only used for type hints
+  from libcloudforensics.providers.aws.internal import account  # pylint: disable=cyclic-import
+
+logging_utils.SetUpLogger(__name__)
+logger = logging_utils.GetLogger(__name__)
+
+IAM_POLICY_DIR = './iampolicies'
+
+# Policy allowing an instance to image a volume
+EBS_COPY_POLICY_DOC = 'ebs_copy_to_s3_policy.json'
+
+#Policy doc to allow EC2 to assume the role. Necessary for instance profiles
+EC2_ASSUME_ROLE_POLICY_DOC = 'ec2_assume_role_policy.json'
+
+
+class IAM:
+  """Class that represents AWS IAM services"""
+  def __init__(self,
+               aws_account: 'account.AWSAccount') -> None:
+    """Initialize the AWS IAM client object.
+
+    Args:
+      aws_account (AWSAccount): An AWS account object.
+    """
+    self.aws_account = aws_account
+    self.client = self.aws_account.ClientApi(common.IAM_SERVICE)
+
+
+  def CheckInstanceProfileExists(self, profile_name: str) -> bool:
+    """Check if an instance role exists.
+
+    Args:
+      profile_name (str): Instance profile name.
+    Returns:
+      bool: True if the Instance Profile exists, false otherwise.
+    """
+    try:
+      self.client.get_instance_profile(InstanceProfileName=profile_name)
+      return True
+    except self.client.exceptions.NoSuchEntityException:
+      return False
+
+  def CreatePolicy(self, name: str, policy_doc: str) -> str:
+    """Creates an IAM policy using the name and policy doc passed in.
+    If the policy exists already, return the Arn of the existing policy.
+
+    Args:
+      name: Name for the policy
+      policy_doc: IAM Policy document as a json string.
+    Returns:
+      str: Arn of the policy.
+    """
+    logger.info("Creating IAM policy {0:s}".format(name))
+    try:
+      policy = self.client.create_policy(
+        PolicyName=name, PolicyDocument=policy_doc)
+      return str(policy['Policy']['Arn'])
+    except self.client.exceptions.EntityAlreadyExistsException as exception:
+      logger.info("Policy exists already, using existing")
+      policies = self.client.list_policies(Scope='Local')
+
+      while True:
```

Comment:
Shouldn't be. We know the policy with that name exists, since to reach this code, the creation must have failed due to the name existing already. At this point we are paging through the results looking for the entry that exists already. If it isn't foind by the time we reach the last page of results (`if not policies['IsTruncated']:`) then we throw an exception.